File: game/systems.cpp

#include "systems.hpp"
#include "components.hpp"
#include<algorithm>



using namespace std;



void game_board::change_ownership(Territory *territory, Player*to) {
	/*
	Player * from = territory->owner;
    territory->owner = to;
    to->territories.push_back(territory);

    from->territories.erase(
        remove_if(begin(from->territories), end(from->territories),
    [&](Territory* t) {
        return t==territory;
    }),
    end(from->territories));
	
	*/
}


// Same as above but with pass-by-reference
void game_board::change_ownership(Territory &territory, Player &to) {
	Player &  from = *territory.owner;	
    territory.owner = &to;
    to.territories.push_back(territory);

    from.territories.erase(
        remove_if(begin(from.territories), end(from.territories),
    [&](Territory t) {
        return t.name==territory.name;
    }),
    end(from.territories));
}

File: Go/game/systems.go
package game

func ChangeOwnership(territory *Territory, to *Player) {
	from := territory.Owner
	territory.Owner = to
	to.Territories = append(to.Territories, territory)

	newTerritories := make([]*Territory, 0, len(from.Territories)-1)
	for _, t := range from.Territories {
		if t != territory {
			newTerritories = append(newTerritories, t)
		}
	}
	from.Territories = newTerritories
}